### ISSUE TITLE
[Compiler Bug]: Cannot alter scroll values on a element -  Mutating a value returned from 'useState'

### ISSUE THREAD

**Issue by lukpsaxo** [facebook/react#31483]
### What kind of issue is this?

- [X] React Compiler core (the JS output is incorrect, or your app works incorrectly after optimization)
- [ ] babel-plugin-react-compiler (build issue installing or using the Babel plugin)
- [ ] eslint-plugin-react-compiler (build issue installing or using the eslint plugin)
- [ ] react-compiler-healthcheck (build issue installing or using the healthcheck script)

### Link to repro

https://playground.react.dev/#N4Igzg9grgTgxgUxALhAMygOzgFwJYSYAEAggBTBFhwwQA2dAChGEQL4CURwAOsUXEJgcRANoI6AGioIcAUToBdIgF4iAJQQBDXADooYBAGUcWnAjIcA3Hz5EiBhABktAT2jy0aBLjKXVAHzcRHb2RBK61LQMTghoImpR9EwsNvxs0uJSVDTJzGCKHLb8MLKwxAA8ACZ4AG5EpWgqwIbydGwBFQD0NbUBaWwgbEA

### Repro steps

Sometimes we need to "mutate" a html element, e.g. when dealing with scrolling.

But with this code we get an error:
> Mutating a value returned from 'useState()', which should not be mutated. Use the setter function to update instead

### How often does this bug happen?

Every time

### What version of React are you using?

18.3.1

### What version of React Compiler are you using?

19.0.0-beta-a7bf2bd-20241110

**Comment by MonstraG** [comment]
Hit the same problem with media tracks:
```tsx
export const VideoButton: FC = () => {
	const localVideoTrack: MediaStreamTrack | undefined = useLocalMediaContext();

	const handleClick = useCallback(() => {
		if (localVideoTrack == null) {
			return;
		}
		// eslint-disable-next-line react-compiler/react-compiler -- https://github.com/facebook/react/issues/31483
		localVideoTrack.enabled = !localVideoTrack.enabled;
	}, [localVideoTrack]);

	return (
		<Button onClick={handleClick}>
			{localVideoTrack?.enabled ? <VideocamIcon /> : <VideocamOffIcon />}
		</Button>
	);
};
```
```
"eslint-plugin-react-compiler": "19.0.0-beta-e993439-20250405",
```

**Comment by josephsavona** [comment]
@lukpsaxo If you're trying to mutate a DOM element, the supported pattern in React is to use a ref: https://react.dev/learn/referencing-values-with-refs#refs-and-the-dom 

Here's an example: https://playground.react.dev/#N4Igzg9grgTgxgUxALhAMygOzgFwJYSYAEAggBTBFhwwQA2dAChGEQL4CURwAOsUXEJgcRGAjREAvEQBKCAIa4AdFDAI5aMhwDcfIkT1FVCADLyAntBwBRNGgS4yWqQD5uB-vrFolcWGMwcJWpaBhNxEWkQ+iYWXX42ABoiAG0EOmTohmYwAF0OPkMxHFhiAB4AEzwAN1FxSWBvNhcygHoq6pd4thA2IA

**Comment by josephsavona** [comment]
@MonstraG Yup, as you noted React's rules prevent modifying hook return values. The intuition here is that this value may be memoized, and mutating it would produce undefined behavior.

The correct approach is as you noted — use state and pass a new state when you need to update.

**Comment by Trickskeee3013** [comment]
This is happening to me and the space is meshing itself together .

**Comment by sofiamiller350** [comment]
I’m testing smart contract interactions on https://duckybsc.xyz but need sоme tokens. Can someone help me out with a few? 0x2143F7146F0AadC0F9d85ea98F23273Da0e002Ab
@josephsavona @MonstraG @lukpsaxo @Trickskeee3013 @sunny

**Comment by stuartkeith** [comment]
> [@lukpsaxo](https://github.com/lukpsaxo) If you're trying to mutate a DOM element, the supported pattern in React is to use a ref: https://react.dev/learn/referencing-values-with-refs#refs-and-the-dom
> 
> Here's an example: https://playground.react.dev/#N4Igzg9grgTgxgUxALhAMygOzgFwJYSYAEAggBTBFhwwQA2dAChGEQL4CURwAOsUXEJgcRGAjREAvEQBKCAIa4AdFDAI5aMhwDcfIkT1FVCADLyAntBwBRNGgS4yWqQD5uB-vrFolcWGMwcJWpaBhNxEWkQ+iYWXX42ABoiAG1ohmYwAF0OPkMxHFhiAB4AEzwAN1FxSWBvNhdigHpyipd4thA2IA

Just following on from this - sometimes storing a ref in state is useful if the ref itself is expected to change. Example:

```
function A({ toggle }) {
  const [el, setEl] = React.useState();

  useLayoutEffect(() => { 
    if (!el) return

    // we don't do this when setting the ref as we want to clean it up when the ref changes,
    // and we're using React 18 that doesn't support this in the ref callback
    el.x = 123
    
    return () => {
       el.x = null;
    }
  }, [el])

  return toggle ? <div ref={setEl}></div> : null;
}
```

With `useRef`, there would be no linting error, but the effect wouldn't have `el` as a dependency, and then it won't re-run when the ref changes.

I'm having trouble finding anyone discussing this one way or the other - if the state is storing a DOM element or other mutable source, and that state is never read in the render itself, does mutating it actually cause any problems with the compiler? Or is it a false positive because it can't tell the difference between that and `useState({ x: 123 })`?

**Comment by josephsavona** [comment]
If you use a ref callback in 18, the callback will be invoked with null when it's unmounted. If you need to keep state between the init and teardown of that ref, the callback function can use another ref to store that state:

```js
const ref = useRef(...);
<div ref={el => {
  if (el == null) {
    // teardown ref.current state
  } else {
    ref.current = init(el);
  }
})
```

Obviously this is easier in 19, but in general you shouldn't need to store a ref in state just to get the desired behavior. I'm gonna close the issue since there isn't a clear action item, but happy to continue discussing.
